src/dbnode/ts/write_batch.go

```diff
@@ -102,7 +107,22 @@ func (b *writeBatch) SetSkipWrite(idx int) {
 	b.writes[idx].SkipWrite = true
 }
 
+func (b *writeBatch) SetFinalizeAnnotationFn(f FinalizeAnnotationFn) {
+	b.finalizeAnnotationFn = f
+}
+
 func (b *writeBatch) Finalize() {
+	if b.finalizeAnnotationFn != nil {
+		for _, write := range b.writes {
+			annotation := write.Write.Annotation
+			if annotation == nil {
+				continue
+			}
+
+			b.finalizeAnnotationFn(annotation)
+		}
+	}
+
 	b.ns = nil
 	b.writes = b.writes[:0]
```

Comment:
unrelated to your changes, but this needs to nil set the writes too or it's leaking. mind adding the following while you're here –
```
var zeroedWrite BatchWrite
for i := range b.writes {
  b.writes[i] = zeroedWrite
}
```